src/management-system-v2/components/process-copy.tsx

```diff
@@ -0,0 +1,344 @@
+import { Button, Checkbox, Collapse, Input, Modal, Spin, Switch, Tooltip } from 'antd';
+import React, { Dispatch, FC, Key, SetStateAction, useCallback, useEffect, useState } from 'react';
+import styles from './process-copy.module.scss';
+import TextArea from 'antd/es/input/TextArea';
+import { useUserPreferences } from '@/lib/user-preferences';
+import { addUserPreference } from '@/lib/utils';
+import { useGetAsset, usePostAsset } from '@/lib/fetch-data';
+import type { CollapseProps } from 'antd';
+import {
+  generateDefinitionsId,
+  setDefinitionsId,
+  setDefinitionsName,
+  setDefinitionsVersionInformation,
+  setTargetNamespace,
+  addDocumentation,
+  getDefinitionsId,
+} from '@proceed/bpmn-helper';
+import { fetchProcessVersionBpmn } from '@/lib/process-queries';
+import { LoadingOutlined } from '@ant-design/icons';
+
+type ProcessCopyModalType = {
+  setCopyProcessIds: Dispatch<SetStateAction<string[]>> | Dispatch<SetStateAction<Key[]>>;
+  processKeys: React.Key[] | String[];
+  setSelection: Dispatch<SetStateAction<string[]>> | Dispatch<SetStateAction<Key[]>>;
+};
+type CopyProcessType = {
+  bpmn: string;
+  newName?: string;
+  newDescription?: string;
+  oldId: string | Key;
+};
+
+const sliceTitleLength = (title: String) => {
+  return title.length > 50 ? `${title.slice(0, 51)} ...` : title;
+};
+
+const ProcessCopyModal: FC<ProcessCopyModalType> = ({
+  setCopyProcessIds,
+  processKeys,
+  setSelection,
+}) => {
+  const { mutateAsync: addProcess } = usePostAsset('/process', {
+    onError: async (error, variables) => {
+      const id = await getDefinitionsId(variables.body.bpmn);
+      setFailed((prev) => [...prev, id as string]);
+    },
+    onSuccess: async (data) => {
+      const id = JSON.parse(data).definitionId;
+      setSelection((prev) => [...prev, id as string]);
+      setSuccessful((prev) => [...prev, id as string]);
+      //   setSelection((prev) => [...prev, data.definitionId as string]);
+      //   setSuccessful((prev) => [...prev, data.definitionId as string]);
+    },
+  });
+
+  const [successful, setSuccessful] = useState<string[]>([]);
+  const [failed, setFailed] = useState<string[]>([]);
+
+  const [loading, setLoading] = useState(false);
+
+  const { preferences, addPreferences } = useUserPreferences();
+
+  const { 'process-copy-modal-accordion': isAccordion, 'ask-before-copying': copyModalPreference } =
+    preferences;
+
+  const { data, refetch: refreshData } = useGetAsset('/process', {
+    params: {
+      query: { noBpmn: true },
+    },
+  });
+
+  const [bluePrintForProcesses, setBluePrintForProcesses] = useState(
```

Comment:
Not sure if that is actually the case. 

The blue prints for the copies are actually not derivable from the selected process keys, as their state changes independently from the processKeys. While their initial state is derived from them and the data of the processes, their state changes with user-input (see Input and TextArea) - even though the processKeys and the process-data are (possibly) still in the same state.

For usage of useMemo I would need another state that depicts the input changes of the user, that should be applied on the blue-prints (dependency). But that wouldn't cause any less re-renders, since the UI should change accordingly to the users input. The blue-print would therefore change just as often.

Reply:
If understand, it is said, that useEffect should only be used when synchronizing with an external system https://react.dev/learn/you-might-not-need-an-effect, but in this case i couldn't think of anything better. Maybe you could use the selected row elements from the ant design table `selectedRows`, but i don't think it makes a difference.